tests/unit/test_charm.py

```diff
@@ -101,14 +140,88 @@ def test__on_image_relation_changed(
     """
     arrange: given monkeypatched builder, openstack manager, image_observer.
     act: when _on_image_relation_changed is called.
-    assert: charm is in active status.
+    assert: charm is in active status and run for the particular related unit is called.
     """
-    monkeypatch.setattr(state.BuilderConfig, "from_charm", MagicMock())
+    image_config = state.ImageConfig(
+        arch=state.Arch.ARM64,
+        bases=(state.BaseImage.FOCAL,),
+        runner_version="",
+        script_url=None,
+        script_secrets=dict(),
+    )
+    monkeypatch.setattr(
+        state.BuilderConfig,
+        "from_charm",
+        MagicMock(return_value=MagicMock(image_config=image_config)),
+    )
     monkeypatch.setattr(proxy, "configure_aproxy", MagicMock())
     monkeypatch.setattr(builder, "install_clouds_yaml", MagicMock())
     monkeypatch.setattr(builder, "run", MagicMock())
```

Comment:
It seems the monkeypatch are the same as line 64 - 71.
Maybe this should be a fixture.